### PR TITLE
Asn1 integer encoding

### PR DESCRIPTION
**Pull request by bjwtaylor**
## Description

Introduce a new function for asn1 encoding of mpi integers. (initial architectural review).

## PR checklist

- [ ] **changelog** TBC
- [ ] **framework PR** not required
- [ ] **mbedtls development PR** not required because: No changes
- [ ] **mbedtls 3.6 PR** not required because: No backports
- **tests**  provided

### CODE REVIEW

File: drivers/builtin/src/asn1write.c [hunk 1]
```diff
@@ -434,4 +436,52 @@ mbedtls_asn1_named_data *mbedtls_asn1_store_named_data(
...
+    }
+
+    mbedtls_mpi_init(&mpi_to_encode);
+    memset(start, 0, buffer_length);
+
+    ret = mbedtls_mpi_read_binary(&mpi_to_encode, integer, integer_length); //read the input string into the mpi
+    if(ret !=0){
+        return ret;
+    }
+    result_size = mbedtls_asn1_write_mpi(p, start, &mpi_to_encode); //encode the mpi into asn1
```
Comment (gilles-peskine-arm) [inline drivers/builtin/src/asn1write.c:483]: Although we can use the MPI function internally for now, I would prefer to make this the primary function, and change `write_mpi` to call it. That way the new function would benefit from the existing tests and it would directly be more future-looking, and it would do no heap allocation.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
The design is on the right track, but to do a better review I'd need to see the new function's documentation. In particular I don't understand the role of `buffer_length`: it seems redundant with `integer_length`?

File: drivers/builtin/src/asn1write.c [hunk 2]
```diff
@@ -434,4 +436,52 @@ mbedtls_asn1_named_data *mbedtls_asn1_store_named_data(
 
     return cur;
 }
+
+size_t mbedtls_get_asn1_integer_frame_length(size_t integer_length){
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply drivers/builtin/src/asn1write.c:440]: Lengths that fit in a single octet (i.e. up to 2^7-1 content octets, i.e. 2^10-1 bits) are definitely not enough. Lengths encoded as two octets would be enough in practice. But we should just let `mbedtls_asn1_write_len` do its job.

File: drivers/builtin/src/asn1write.c [hunk 3]
```diff
@@ -434,4 +436,52 @@ mbedtls_asn1_named_data *mbedtls_asn1_store_named_data(
...
+}
+
+/*
+ * to test
+ * buffer too small
+ * buffer to big
+ * different integer length to buffer size
+ * Oversize integer length
+ */
+int mbedtls_asn1_write_integer(unsigned char **p, unsigned char *start, const unsigned char *integer, size_t integer_length, size_t buffer_length) {
```
[the remark this answers is not in the document]
Comment (bjwtaylor) [reply drivers/builtin/src/asn1write.c]: It's also redundant with start and p, I've removed it.

File: include/mbedtls/asn1write.h [hunk 4]
```diff
@@ -381,6 +362,31 @@ mbedtls_asn1_named_data *mbedtls_asn1_store_named_data(mbedtls_asn1_named_data *
                                                        const unsigned char *val,
                                                        size_t val_len);
 
+/**
+ * \brief                  Encode an integer into ASN1 and write it to a buffer.
+ *                         The function takes a big endian bignum buffer that
+ *                         begins at start and ends at *p and encodes it into
+ *                         a asn1 frame.
+ *                         Warning: Destructive behaviour for the buffer data!
```
Comment (gilles-peskine-arm) [inline include/mbedtls/asn1write.h:370]: I don't understand this warning. It seems to contradict the `const` annotation on `integer`.

I would strongly expect `integer` to be const. It's an input, not a working buffer!

File: include/mbedtls/asn1write.h [hunk 5]
```diff
@@ -381,6 +362,31 @@ mbedtls_asn1_named_data *mbedtls_asn1_store_named_data(mbedtls_asn1_named_data *
...
+ *
+ * \param p                The reference to the current position pointer.
+ * \param start            The start of the buffer, for bounds-checking.
+ * \param integer          The start of the integer buffer.
+ * \param integer_length   The size of the integer in the buffer.
+ * \param sign             The signess of the integer in the buffer 1 for
+ *                         positive and -1 for negative.
```
Comment (gilles-peskine-arm) [inline include/mbedtls/asn1write.h:377]: We don't want negative integers. The fact that they're currently partly supported is a historical artifact, and even now they don't work correctly. So this parameter should not exist.

It's possible that we still need support for negative integers internally. If so, that should be handled by an internal function. But I think we don't need that — we still need support for negative `mbedtls_mpi` objects, but I don't think we ever need to format one as ASN.1.

If it's just for unit tests, we should just remove the unit tests with negative values.

File: include/mbedtls/asn1write.h [hunk 6]
```diff
@@ -381,6 +362,31 @@ mbedtls_asn1_named_data *mbedtls_asn1_store_named_data(mbedtls_asn1_named_data *
...
+ * \brief                  Encode an integer into ASN1 and write it to a buffer.
+ *                         The function takes a big endian bignum buffer that
+ *                         begins at start and ends at *p and encodes it into
+ *                         a asn1 frame.
+ *                         Warning: Destructive behaviour for the buffer data!
+ *
+ * \param p                The reference to the current position pointer.
+ * \param start            The start of the buffer, for bounds-checking.
+ * \param integer          The start of the integer buffer.
+ * \param integer_length   The size of the integer in the buffer.
```
Comment (gilles-peskine-arm) [inline include/mbedtls/asn1write.h:375]: “Size” can be ambiguous. (Number of bits? Number of bytes? Number of words? ...) It's better to explicitly say “number of bytes”.

File: include/mbedtls/asn1write.h [hunk 7]
```diff
@@ -381,6 +362,31 @@ mbedtls_asn1_named_data *mbedtls_asn1_store_named_data(mbedtls_asn1_named_data *
...
+/**
+ * \brief                  Encode an integer into ASN1 and write it to a buffer.
+ *                         The function takes a big endian bignum buffer that
+ *                         begins at start and ends at *p and encodes it into
+ *                         a asn1 frame.
+ *                         Warning: Destructive behaviour for the buffer data!
+ *
+ * \param p                The reference to the current position pointer.
+ * \param start            The start of the buffer, for bounds-checking.
+ * \param integer          The start of the integer buffer.
```
Comment (gilles-peskine-arm) [inline include/mbedtls/asn1write.h:374]: It would be nice to recall the buffer representation here. In fact, if it's only in one place, I think it makes more sense for it to be in the parameter description than in the function's general description.

Comment (gilles-peskine-arm) [inline include/mbedtls/asn1write.h:374]: Is it ok if `integer` overlaps with the output buffer? That can be useful in typical cases that we'll encounter when printing out a bignum. For that we first convert the bignum to a big-endian byte array, then convert that to ASN.1. If we can do the second step in place, then we don't need to allocate an intermediate step.

We may not want to promise that it works if it's too difficult to implement. We do have to be careful that the ASN.1 representation is usually a few bytes larger than the byte array (because there's 2 or more bytes of overhead for the ASN.1 type, ASN.1 length and the possible additional byte for the ASN.1 sign bit), but it can also be smaller if the byte array has leading zeros.

**Comment by gilles-peskine-arm** [comment]
https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/184 is merged, so please rebase on top of `development`.

**Comment by mpg** [comment]
Note: this PR is based on an outdated version of #184 which has been rebased in the meantime, so the commits from #184 has different IDs here. In case git has trouble with that in the rebase, I think we can use the author field to identify commits that are new in this PR: those from Ben, ie starting with "Move mbedtls_asn1_write_mpi to bignum.h as it is now private" - all the previous ones can be discarded.

**Comment by davidhorstmann-arm** [comment]
Clean rebase, no conflicts. We'll have to see what the CI says.

**Comment by valeriosetti** [comment]
> Clean rebase, no conflicts. We'll have to see what the CI says.

A part from Windows related failures which are all related to the CI, there are some code style fixes to be done unfortunately.

**Comment by mpg** [comment]
On the internal CI, the only failures are indeed `check_files` and `check_code_style`.

I will point out those are easy and pretty quick to run locally as part of a pre-commit or pre-push hook (on my 8-years old laptop, `check_files.py` is about 5s, `code_style.py` a bit more than 1 min if checking all files, but only seconds if you run it as part of a pre-commit hook where in only checks files that have changed), so I find it a bit sad that we have to spend so many push-wait-for-ci-push-again cycles just on that (this is far from being the first PR in this case).

File: include/mbedtls/asn1write.h (hunk 7 above)
Comment (davidhorstmann-arm) [reply include/mbedtls/asn1write.h]: Currently overlap is not handled, but it should be possible using `memmove()`. I'll have a careful think about it tomorrow and see if it is easy to do.

File: include/mbedtls/asn1write.h [hunk 8]
```diff
@@ -381,6 +362,30 @@ mbedtls_asn1_named_data *mbedtls_asn1_store_named_data(mbedtls_asn1_named_data *
...
+ * \note                   This function works backwards within the data buffer.
+ *
+ * \param p                The reference to the current position pointer.
+ * \param start            The start of the buffer, for bounds-checking.
+ * \param integer          Pointer to a big-endian representation of an integer.
+ * \param integer_length   The number of bytes in the integer buffer.
+ *
+ * \return                 An integer number of bytes written on success.
+ * \return                 An appropriate error code on failure.
+ * \return                 On success p will be set to point to the start of
```
Comment (gilles-peskine-arm) [inline include/mbedtls/asn1write.h:380]: Minor:
```suggestion
 * \return                 On success \c *p will be set to point to the start of
```

File: tests/suites/test_suite_bignum_core.misc.data [hunk 9]
```diff
@@ -538,3 +538,33 @@ asn1_get_and_parse_integer:"028201000123456789abcdef0123456789abcdef0123456789ab
 
 Parse raw ASN.1 integer, correct encoding, more data than needed
 asn1_get_and_parse_integer:"02010707":0:"07"
+
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_bignum_core.misc.data:541]: I'm struggling to understand the test coverage.

Why isn't `mbedtls_asn1_write_integer()` tested in `test_suite_asn1write`?

I would have some positive tests with an input and the known corresponding output, but I don't see any. `encode_decode_asn1_bignum` does write-then-parse checking, but that could miss bugs if write() does something wrong and out parse() happens to be sloppy and compensates (which is a common design defect in such pairs of functions).

The names of `size_checks`, `null_pointer_checks` and `endian_check` should reflect what they're about. Also, I don't understand what the test goals of `size_checks` and `endian_check` are. For `null_pointer_checks`, I sort of understand, except that it's asserting some inputs that I would have expected to lead to a null pointer dereference rather than returning an error code (null pointer with a non-null size, or null pointer that isn't nullable).

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I partially reviewed c489f5890c20ad9555d518ec6906b826a5c73002, only for the interface (changes to header files) and the test coverage.

* Interface: LGTM
* Test coverage: I don't really understand what we're testing.

One pending comment from previous reviews:

* https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/289#discussion_r2140508935 — this would be nice to have, but I'm perfectly happy to file it as a later improvement.

I don't mind rewriting the commit history at this point, and in fact I would be somewhat favorable to simplifying the history (50 commits is a lot for a patch of this size).

File: drivers/builtin/src/asn1write.c [hunk 10]
```diff
@@ -434,4 +435,77 @@ mbedtls_asn1_named_data *mbedtls_asn1_store_named_data(
...
+
+    integer_start = (unsigned char *) integer + number_of_leading_zeros;
+
+    integer_length -= number_of_leading_zeros;
+
+    if (output_buffer_size < integer_length) {
+        return MBEDTLS_ERR_ASN1_BUF_TOO_SMALL;//TC3 buffer less than integer size.
+    }
+
+    if (start != *p) {
```
[the remark this answers is not in the document]
Comment (davidhorstmann-arm) [reply drivers/builtin/src/asn1write.c:469]: The problem here is that in this case we might have `start == *p == NULL`, in which case `memset()` complains (even for a zero-length buffer).

File: tests/suites/test_suite_asn1write.function [hunk 11]
```diff
@@ -599,3 +599,48 @@ exit:
...
+    unsigned char *output = NULL;
+    unsigned char *p = NULL;
+    size_t output_length = expected_output->len;
+    int ret;
+
+    /* Test with exactly the right output buffer length */
+    TEST_CALLOC(output, output_length);
+    p = output + output_length;
+
+    ret = mbedtls_asn1_write_integer(&p, output, integer->x, integer->len);
```
[the remark this answers is not in the document]
Comment (davidhorstmann-arm) [reply tests/suites/test_suite_asn1write.function:615]: I've added reallocation anyway, with a comment explaining that this helps for zero-length inputs.

File: drivers/builtin/src/asn1write.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (davidhorstmann-arm) [reply drivers/builtin/src/asn1write.c:469]: Hmm. In that case I will remove the special case and test only output buffer lengths from 1 byte upwards (since that will lead to `NULL` dereference).

Would you like me to add a test with a non-`NULL`, zero-length output buffer?

[the remark this answers is not in the document]
Comment (davidhorstmann-arm) [reply drivers/builtin/src/asn1write.c:469]: I have added a test for zero-length non-`NULL` output buffers just for thoroughness. It won't allow ASan to detect overflow but it should make sure we are doing roughly the 'right' thing.

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply drivers/builtin/src/asn1write.c]: I filed https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/340 for using `mbedtls_asn1_write_raw_buffer()`.

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM at bc7deb5af1c4a9917b3d949f65c0ee281f1b0508

This will need a rebase on top of a very recent `development` to pass the GCC 15 test. I wouldn't mind simplifying the history at that point. But please do the history rewrite and the base change in separate rebases, so that they can be reviewed separately.

**Review by bjwtaylor — COMMENTED** [review]
LGTM

**Review by gilles-peskine-arm — APPROVED** [review]
Rebase to 2a3e2c5ea053c14b745dbdf41f609b1edc6a72fa approved

**Comment by davidhorstmann-arm** [comment]
Second rebase completed (sorry for the delay). I have squashed some commits together and done some merging so that there are only 45 commits total (rather than 67).

The commits from https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/289/commits/1a3e2cbe13232ce10bc92b781eb81a6003890c3a onwards I have left unchanged. before that point there were previously 39 commits and there are now 18.

If this is still too much to review commit-by-commit I can do some more rebasing work, let me know.

**Review by gilles-peskine-arm — DISMISSED** [review]
Explicitly approving after the partial history squash from a39d1ea2cb344cf3204b611165171b15cb678053 to 4e726aee1b2855350fb7f062316e9bb52f8e162c

**Review by davidhorstmann-arm — DISMISSED** [review]
Approving on behalf of:

* Myself, for @bjwtaylor's commits
* @bjwtaylor, for my commits

**Comment by mpg** [comment]
I think this needs a ChangeLog entry. https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/337/ took care of the migration guide part, but we still need a ChangeLog entry. (We have one of for the parsing side.)

**Comment by davidhorstmann-arm** [comment]
> I think this needs a ChangeLog entry. #337 took care of the migration guide part, but we still need a ChangeLog entry. (We have one of for the parsing side.)

Ah you're right, I'll add that momentarily.

**Comment by bjwtaylor** [comment]
Changelog also looks good.
